experiments/MoarCSets/SInstances.jl

```diff
@@ -0,0 +1,406 @@
+  Ob[Codom[cd_hom_num(CD,hom)]]
+end
+
+struct AttrDesc{CD,Data,Attr,ADom,ACodom}
+  function AttrDesc{CD,Data,Attr,ADom,ACodom}() where {CD,Data,Attr,ADom,ACodom}
+    new{CD,Data,Attr,ADom,ACodom}()
+  end
+  function AttrDesc(pres::Presentation{Schema})
+    CD = CatDescType(pres)
+    datas, attrs = generators(pres, :Data), generators(pres,:Attr)
+    data_syms, attr_syms = nameof.(datas), nameof.(attrs)
+    ob_num = ob -> findfirst(CD.ob .== ob)::Int
+    data_num = ob -> findfirst(data_syms .== ob)::Int
+    new{CD,Tuple(data_syms), Tuple(attr_syms),
+        Tuple(@. ob_num(nameof(dom(attrs)))), Tuple(@. data_num(nameof(codom(attrs))))}()
+  end
+  function AttrDesc(::CatDesc{Ob,Hom,Dom,Codom}) where {Ob,Hom,Dom,Codom}
+    new{CatDesc{Ob,Hom,Dom,Codom},(),(),(),()}
+  end
+end
+
+AttrDescType(pres::Presentation{Schema}) = typeof(AttrDesc(pres))
+
+function Base.getproperty(AD::Type{T}, i::Symbol) where
+  {CD,Data,Attr,ADom,ACodom,T <: AttrDesc{CD,Data,Attr,ADom,ACodom}}
+  @match i begin
+    :cd => CD
+    :data => Data
+    :attr => Attr
+    :adom => ADom
+    :acodom => ACodom
+    _ => invoke(Base.getproperty, Tuple{Type,Symbol}, AD, i)
+  end
+end
+
+function ad_data_num(AD::Type{T}, data::Symbol) where
+  {CD,Data,Attr,ADom,ACodom,T <: AttrDesc{CD,Data,Attr,ADom,ACodom}}
+  findfirst(Data .== data)::Int
+end
+
+function ad_attr_num(AD::Type{T}, attr::Symbol) where
+  {CD,Data,Attr,ADom,ACodom,T <: AttrDesc{CD,Data,Attr,ADom,ACodom}}
+  findfirst(Attr .== attr)::Int
+end
+
+function ad_dom(AD::Type{T}, attr::Symbol) where
+  {CD,Data,Attr,ADom,ACodom,T <: AttrDesc{CD,Data,Attr,ADom,ACodom}}
+  CD.ob[ADom[ad_attr_num(AD,attr)]]
+end
+
+function ad_codom(AD::Type{T}, attr::Symbol) where
+  {CD,Data,Attr,ADom,ACodom,T <: AttrDesc{CD,Data,Attr,ADom,ACodom}}
+  Data[ACodom[ad_attr_num(AD,attr)]]
+end
+
+function make_index(CD::Type{<:CatDesc},AD::Type{<:AttrDesc},Ts::Type{<:Tuple},Idxed::Tuple)
+    function make_idx(name)
+      if name ∈ CD.hom
+        Vector{Vector{Int}}()
+      elseif name ∈ AD.attr
+        Dict{Ts.parameters[ad_codom(AD,name)],Vector{Int}}()
+      else
+        error("invalid index parameter")
+      end
+    end
+  NamedTuple{Idxed}(Tuple(make_idx(name) for name in Idxed))
+end
+
+function table_desc(::Type{CatDesc{Ob,Hom,Dom,Codom}},
+                    ::Type{AttrDesc{CatDesc{Ob,Hom,Dom,Codom},Data,Attr,ADom,ACodom}},
+                    ts::Tuple) where {Ob,Hom,Dom,Codom,Data,Attr,ADom,ACodom}
+  cols = [Tuple{Symbol,DataType}[] for ob in Ob]
+  for i in 1:length(Hom)
+    push!(cols[Dom[i]], (Hom[i],Int))
+  end
+  for i in 1:length(Attr)
+    push!(cols[ADom[i]], (Attr[i],ts[ACodom[i]]))
+  end
+  [(col_names = first.(cols[i]), col_types = last.(cols[i])) for i in 1:length(Ob)]
+end
+
+abstract type AbstractSInstance{CD,AD,Ts} end
+
+struct SInstance{CD <: CatDesc, AD <: AttrDesc{CD}, Ts <: Tuple, Idxed,
```

Comment:
Not sure what would be the best name, but I'm thinking that `SchemaInst` would be more memorable than `SInstance`. Or we could really go for it and use `DBInstance`, in which case we might also rename `Schema` to `DBSchema`. What do you think?

Reply:
I don't like `DBInstance`; I don't want people thinking of these things as databases because people think of databases as storing data long term, and normally you only have one database. `SchemaInst` is nice, if a bit long and `Inst` is kind of hard to pronounce. I have to say, `CSet` is a really nice name, but this isn't quite it...

We could call it `DCSet` for "decorated CSet"? Let me keep thinking.

Reply:
Not sure about the general name, but at the very least we should have type aliases

```
const AbstractCSet{CD} = AbstractSchemaInst{CD,(),()}
const CSet{CD} = SchemaInst{CD,(),()}
```

but these statements are not misleading.

Reply:
We might want to add the index parameters to `CSet`, but yeah, I like that a lot, and that allows us to write all of the limits and colimits for `CSet`s!